modules/operations/cycle_highway_tag.js

```diff
@@ -6,31 +6,62 @@ import { KeyOperationBehavior } from '../behaviors/KeyOperationBehavior';
 let _wasSelectedIDs = [];
 let _wasPresetIDs = [];
 
-
 export function operationCycleHighwayTag(context, selectedIDs) {
   const editor = context.systems.editor;
   const graph = editor.staging.graph;
   const l10n = context.systems.l10n;
   const presets = context.systems.presets;
 
   // Allow cycling through lines that match these presets
-  const allowPresetRegex = [
+  const allowHighwayPresetRegex = [
     /^highway\/(motorway|trunk|primary|secondary|tertiary|unclassified|residential|living_street|service|track)/,
-    /^line$/
+    /^line$/,
   ];
 
-  const defaultPresetIDs = [
+  const defaultHighwayPresetIDs = [
     'highway/residential',
     'highway/service',
     'highway/track',
     'highway/unclassified',
     'highway/tertiary',
-    'line'
+    'line',
+  ];
+
+  // Allow cycling through crossings that match these presets
+  const allowCrossingPresetRegex = [
+    /^crossing\/(unmarked|marked|marked:zebra|marked:lines|marked:dashes|marked:ladder|marked:dots|marked:ladder:skewed)/,
+  ];
+
+  const defaultCrossingPresetIDs = [
+    'crossing/unmarked;crossing:markings=no',
+    'crossing/marked;crossing:markings=yes',
+    'crossing/marked;crossing:markings=zebra',
+    'crossing/marked;crossing:markings=lines',
+    'crossing/marked;crossing:markings=ladder',
+    'crossing/marked;crossing:markings=dashes',
+    'crossing/marked;crossing:markings=dots',
+    'crossing/marked;crossing:markings=ladder:skewed',
   ];
 
-  // same selection as before?
-  const isSameSelection = utilArrayIdentical(selectedIDs, _wasSelectedIDs);
-  const presetIDs = new Set(isSameSelection ? _wasPresetIDs : defaultPresetIDs);
+  // Check if selection is highway or crossing
+  const isCrosswalkSelection = selectedIDs.some((id) => id.includes('crosswalk'));
+  const isHighwaySelection = selectedIDs.some((id) => id.includes('highway'));
+
+  // Define the preset IDs based on the selection type
+  let presetIDs;
+
+  // Declare isSameSelection here
+  let isSameSelection = utilArrayIdentical(selectedIDs, _wasSelectedIDs);
+
+  if (isCrosswalkSelection) {
+    presetIDs = defaultCrossingPresetIDs;
+  } else if (isHighwaySelection) {
+    presetIDs = isSameSelection ? _wasPresetIDs : defaultHighwayPresetIDs;
+  } else {
+    return;
```

Comment:
Another issue you're probably seeing is stack traces when you select things- the reason for this is that the code you've added is all oustide the operation() method itself, which starts on line 54. 

All the new code that you've written up to this line should go into the operation() method. Try to compare your code structure to the original, unmodified cycle_highway_tags.js structure: There are no 'return true' or 'return false' outside the operation or its methods. 

Think of it this way- the return value of operationCycleHighwayTags() should be an operation, and it should have several methods attached to it like annotation(), disabled(), etc. It should NOT just be a simple true or false value.